File: Java Based Archives/WorkManager/app/src/main/java/com/android/workmanager/MainActivity.java
package com.android.workmanager;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.work.Constraints;
import androidx.work.Data;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;

public class MainActivity extends AppCompatActivity {
    private Button button;
    private TextView textView;
    public static final String KEY_COUNT_VALUE = "key count value";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main); // Set the layout for the activity.

        // Create a Data object to hold key-value pairs.
        Data data = new Data.Builder().putInt(KEY_COUNT_VALUE, 4500).build();

        // Create Constraints for the work request.
        Constraints constraint = new Constraints.Builder()
                .setRequiresCharging(true)
                .build();

        // Create a OneTimeWorkRequest for the DemoWorker class with input data and constraints.
        final OneTimeWorkRequest oneTimeWorkRequest = new OneTimeWorkRequest.Builder(DemoWorker.class)
                //.setConstraints(constraint)
                .setInputData(data)
                .build();

        // Find and initialize the Button and TextView from the layout.
        button = findViewById(R.id.button);
        textView = findViewById(R.id.tvStatus);

        // Set a click listener for the Button.
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Enqueue the work request to the WorkManager when the Button is clicked.
                WorkManager.getInstance().enqueue(oneTimeWorkRequest);
            }
        });

        // Observe the WorkInfo associated with the oneTimeWorkRequest.
        WorkManager.getInstance().getWorkInfoByIdLiveData(oneTimeWorkRequest.getId())
                .observe(this, new Observer<WorkInfo>() {
                    @Override
                    public void onChanged(WorkInfo workInfo) {
                        if (workInfo != null) {
                            // Update the TextView with the current state of the work request.
                            textView.setText(workInfo.getState().name());
                        }
                        if (workInfo.getState().isFinished()) {
                            // If the work request is finished, retrieve output data and display a Toast message.
                            Data data1 = workInfo.getOutputData();
                            String message = data1.getString(DemoWorker.KEY_WORKER);
                            Toast.makeText(getApplicationContext(), message, Toast.LENGTH_LONG).show();
                        }
                    }
                });
    }
}

File: WorkManagerDemo2/app/src/main/java/com/example/workmanagerdemo2/MainActivity.kt
package com.example.workmanagerdemo2

import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.work.Constraints
import androidx.work.Data
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkInfo
import androidx.work.WorkManager
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity() {

    companion object {
        const val KEY_COUNT_VALUE = "key_count"
    }

    private lateinit var textView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Set the layout for this activity using the XML layout file (activity_main.xml)
        setContentView(R.layout.activity_main)

        // Initialize TextView and Button by finding them using their respective IDs
        textView = findViewById(R.id.textView)
        val button = findViewById<Button>(R.id.button)

        // Set an OnClickListener on the Button
        button.setOnClickListener {
            // When the Button is clicked, invoke the setOneTimeWorkRequest function
           // setOneTimeWorkRequest()
            setPeriodicWorkRequest()
        }
    }

    // Function to set up and enqueue a one-time work request
    private fun setOneTimeWorkRequest() {

        // Get an instance of WorkManager
        val workManager: WorkManager = WorkManager.getInstance(applicationContext)

        // Create input data for the work request
        val data: Data = Data.Builder()
            .putInt(KEY_COUNT_VALUE, 125)
            .build()

        // Define constraints for the work request
        val constraints = Constraints.Builder()
            .setRequiresCharging(true)
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        // Create a OneTimeWorkRequest for the UploadWorker class with specified constraints and input data
        val uploadRequest = OneTimeWorkRequest.Builder(UploadWorker::class.java)
            .setConstraints(constraints)
            .setInputData(data)
            .build()

        // Create separate work requests for Filtering, Compression, and Downloading workers
        val filteringRequest: OneTimeWorkRequest = OneTimeWorkRequest.Builder(FilteringWorker::class.java)
            .build()

        val compressingRequest: OneTimeWorkRequest = OneTimeWorkRequest.Builder(CompressionWorker::class.java)
            .build()

        val downloadingWorker: OneTimeWorkRequest = OneTimeWorkRequest.Builder(DownloadingWorker::class.java)
            .build()

        // Create a list to hold parallel work requests
        val parallelWorks: MutableList<OneTimeWorkRequest> = mutableListOf<OneTimeWorkRequest>()

        // Add Filtering and Downloading workers to the parallelWorks list
        parallelWorks.add(downloadingWorker)
        parallelWorks.add(filteringRequest)

        // Enqueue the work requests using WorkManager in the specified order
        workManager
            .beginWith(parallelWorks) // Run Filtering and Downloading workers in parallel
            .then(compressingRequest) // Run Compression worker after the parallel workers
            .then(uploadRequest) // Run Upload worker after the Compression worker
            .enqueue()

        // Observe the WorkInfo for the Upload worker
        workManager.getWorkInfoByIdLiveData(uploadRequest.id)
            .observe(this, Observer<WorkInfo> { workInfo ->
                textView.text = workInfo?.state?.name ?: "WorkInfo is null"
                if (workInfo.state.isFinished) {
                    // When the work is finished, display a Toast with the output message
                    val data = workInfo.outputData
                    val message: String? = data.getString(UploadWorker.KEY_WORKER)
                    Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show()
                }
            })
    }
    private fun setPeriodicWorkRequest(){
        val periodicWorkRequest = PeriodicWorkRequest.Builder(DownloadingWorker::class.java, 15, TimeUnit.MINUTES)
            .build()
        WorkManager.getInstance(applicationContext).enqueue(periodicWorkRequest)
    }
}
